packages/protocol/contracts/stability/GrandaMento.sol

```diff
@@ -0,0 +1,254 @@
+  }
+
+  /**
+   * @notice Used in place of the constructor to allow the contract to be upgradable via proxy.
+   * @param _registry The address of the registry.
+   * @param _spread The spread charged on exchanges.
+   */
+  function initialize(address _registry, uint256 _spread) external initializer {
+    _transferOwnership(msg.sender);
+    setRegistry(_registry);
+    setSpread(_spread);
+  }
+
+  /**
+   * @notice Creates a new exchange proposal and deposits the tokens being sold.
+   * @dev Stable token value amounts are used for the sellAmount, not unit amounts.
+   * @param stableToken The stableToken involved in the exchange.
+   * @param sellAmount The amount of the sell token being sold.
+   * @param sellCelo Whether CELO is being sold.
+   * @return The proposal identifier for the newly created exchange proposal.
+   */
+  function createExchangeProposal(address stableToken, uint256 sellAmount, bool sellCelo)
+    external
+    nonReentrant
+    returns (uint256)
+  {
+    // Require the configurable stableToken max exchange amount to be > 0.
+    // This covers the case where a stableToken has never been explicitly permitted.
+    ExchangeLimits memory exchangeLimits = stableTokenExchangeLimits[stableToken];
+    require(exchangeLimits.maxExchangeAmount > 0, "Max stable token exchange amount must be > 0");
+
+    // Using the current oracle exchange rate, calculate what the buy amount is.
+    // This takes the spread into consideration.
+    uint256 buyAmount = getBuyAmount(stableToken, sellAmount, sellCelo);
+
+    // Ensure that the amount of stableToken being bought or sold is within
+    // the configurable exchange limits.
+    uint256 stableTokenExchangeAmount = sellCelo ? buyAmount : sellAmount;
+    require(
+      stableTokenExchangeAmount <= exchangeLimits.maxExchangeAmount &&
+        stableTokenExchangeAmount >= exchangeLimits.minExchangeAmount,
+      "Stable token exchange amount not within limits"
+    );
+
+    // Deposit the assets being sold.
+    IERC20 sellToken = sellCelo ? getGoldToken() : IERC20(stableToken);
+    require(
+      sellToken.transferFrom(msg.sender, address(this), sellAmount),
+      "Transfer of sell token failed"
+    );
+
+    // Record the proposal.
+    uint256 proposalId = exchangeProposalCount;
+    exchangeProposals[proposalId] = ExchangeProposal({
+      exchanger: msg.sender,
+      stableToken: stableToken, // for stable tokens, is saved in units to deal with demurrage.
+      sellAmount: sellCelo ? sellAmount : IStableToken(stableToken).valueToUnits(sellAmount),
+      buyAmount: buyAmount,
+      approvalTimestamp: 0, // initial value when not approved yet
+      state: ExchangeState.Proposed,
+      sellCelo: sellCelo
+    });
+    exchangeProposalCount = exchangeProposalCount.add(1);
+    // Even if stable tokens are being sold, the sellAmount emitted is the "value."
+    emit ProposedExchange(proposalId, msg.sender, stableToken, sellAmount, buyAmount, sellCelo);
```

Comment:
why as value instead of units?

Reply:
I don't have a _super_ strong opinion on this, but imo the storing of stable token units is an implementation detail and events should expose things that are useful for observing what's happened -- in this case, it's showing that the proposer has deposited exactly `sellAmount` tokens to buy exactly `buyAmount`. I think requiring consumers to process an event and then call `StableToken.unitsToValue` to figure out the amount being sold is slightly confusing